### PR TITLE
Network viewer charts

### LINKED ISSUE #9599: Network Viewer Charts MVP0
##### Feature idea summary

We need to bring the following new charts as MVP0 of the Network Viewer:

 - ebpf.outbound_bytes:
     - `<port|service name>:<protocol>:<ip|hostname>_sent`
     - `<port|service name>:<protocol>:<ip|hostname>_received`
     - `other_sent`
     - `other_received`
   - ebpf.outbound_packets:
      - `<port|service name>:<protocol>:<ip|hostname>_sent`
      - `<port|service name>:<protocol>:<ip|hostname>_received`
     -  `other_sent`
     -  `other_received`
  - ebpf.outbound_retransmits:  `<port>:<protocol>:<ip>`
     -  ebpf.inbound_bytes:
        - `<port|service name>:<protocol>_sent`
        - `<port|service name>:<protocol>_received`
        - `other_sent`
         - `other_received`
  - ebpf.inbound_packets
     - `<port|service name>:<protocol>_sent`
     - `<port|service name>:<protocol>_received`
     - `other_sent`
     - `other_received`

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
Fixes #9599

##### Component Name
Collector
##### Test Plan

1 - Configure your `/etc/netdata/ebpf.conf`
2 - Compile this branch
3 - Start Netdata
More details after the blocker to be merged

### CODE REVIEW

File: collectors/ebpf.plugin/ebpf_socket.c [hunk 1]
```diff
@@ -289,12 +465,713 @@ void ebpf_socket_create_apps_charts(ebpf_module_t *em, struct target *root)
...
+ *
+ * @param sock
+ * @param lvalues
+ */
+static inline void update_socket_data(netdata_socket_t *sock, netdata_socket_t *lvalues)
+{
+    sock->recv_packets += lvalues->recv_packets;
+    sock->sent_packets += lvalues->sent_packets;
+    sock->recv_bytes   += lvalues->recv_bytes;
+    sock->sent_bytes   += lvalues->sent_bytes;
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_socket.c:976]:
```suggestion
    sock->recv_packets = lvalues->recv_packets;
    sock->sent_packets = lvalues->sent_packets;
    sock->recv_bytes   = lvalues->recv_bytes;
    sock->sent_bytes   = lvalues->sent_bytes;
```

File: collectors/ebpf.plugin/ebpf.h [hunk 2]
```diff
@@ -189,10 +189,10 @@ extern void write_end_chart();
 #define EBPF_NETWORK_VIEWER_SECTION "network viewer"
 #define EBPF_SERVICE_NAME_SECTION "service name"
 
-#define EBPF_COMMON_DIMENSION_CALL "Calls"
-#define EBPF_COMMON_DIMENSION_PACKETS "Packets"
+#define EBPF_COMMON_DIMENSION_CALL "calls"
 #define EBPF_COMMON_DIMENSION_BYTESS "bytes/s"
-#define EBPF_COMMON_DIMENSION_DIFFERENCE "Difference"
+#define EBPF_COMMON_DIMENSION_DIFFERENCE "difference"
+#define EBPF_COMMON_DIMENSION_PACKETS "packets"
```
Comment (ilyam8) [inline collectors/ebpf.plugin/ebpf.h:195]: only bytes `/s` ?

**Comment by vlvkobal** [comment]
image
image

**Comment by vlvkobal** [comment]
This is an ssh connection which had been established before Netdata started
```
2020-07-29 12:02:40: ebpf.plugin INFO  : EBPF SOCKET READ : New outbound dimension added: ID = 0, IP = 192.168.0.126, NAME = 192.168.0.126, DIM1 = 49680:TCP:192.168.0.126_recv, DIM2 = 49680:TCP:192.168.0.126_sent, DIM3 = 49680:TCP:192.168.0.126
```
I can see traffic on the chart when I type anything in the terminal.

I start another ssh connection
```
2020-07-29 12:07:35: ebpf.plugin INFO  : EBPF SOCKET READ : New inbound dimension added: ID = 9, IP = 192.168.0.126, NAME = 192.168.0.168, DIM1 = 22:TCP_recv, DIM2 = 22:TCP_sent, DIM3 = 22:TCP
2020-07-29 12:07:35: ebpf.plugin INFO  : EBPF SOCKET READ : New inbound dimension added: ID = 10, IP = 192.168.0.126, NAME = 192.168.0.168, DIM1 = 22:TCP_recv, DIM2 = 22:TCP_sent, DIM3 = 22:TCP
```
The previous connection that was classified at first as outbound was correctly reclassified and a new connection added. Now I can see traffic only for the second shh connection.

**Comment by vlvkobal** [comment]
`resolve hostname ips` configuration option should be `resolve names`. It affects both hostnames and services. 
But I think it's more logical to have separate options:
```
resolve hostnames
resolve service names
```
@dim08 @ilyam8 what do you think?

**Comment by ilyam8** [comment]
> It affects both hostnames and services.
> resolve hostnames

These services are from `/etc/services`, right? Why do we need `resolve hostnames`? I think it should be always `on`. Any use cases when a user want to have it turned off?

**Comment by ilyam8** [comment]
>  classified at first as outbound was correctly

How does classification work for UDP? Is it possible to understand it is rx/tx traffic?

**Comment by vlvkobal** [comment]
> Any use cases when a user want to have it turned off?

Sometimes service names don't help at all, I usually prefer to see port numbers. Moreover, default names can be misleading.

**Comment by ilyam8** [comment]
> Sometimes service names don't help at all, I usually prefer to see port numbers. Moreover, default names can be misleading.

then dont use `services` ?

```
services = !*
ports = ...
```

File: collectors/ebpf.plugin/ebpf.h (hunk 2 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf.h:195]: `bytes/s` are reclassified to `kbyte/s` or another dimension when it i necessary. I think I did not understand your question.

**Comment by amoss** [comment]
> > classified at first as outbound was correctly
> 
> ~How does classification work for UDP? Is it possible to understand it is rx/tx traffic?~
> 
> ok, my question was not about `rx/tx traffic` but inbound/outbound connection and we classify it based on sockets in listen state.

The inbound / outbound refers to which direction the connection was established, and then within both kinds there is rx/tx traffic.

For UDP (and anything else that is connectionless) all tx is outbound, and all rx is inbound.

**Comment by thiagoftsm** [comment]
> image
> image

I missed an dimension update, thank you to remember me this.

**Comment by underhood** [comment]
Screenshot from 2020-07-29 13-56-22

2 questions here:
1. do we have an idea why the chart looks like saw teeth?
2. wouldn't it be better for these chart dimensions to be stacked (instead of they being over each other)?

**Comment by vlvkobal** [comment]
> do we have an idea why the chart looks like saw teeth

I presume it's a Netdata connection so charts are updating every second. I think it's a consequence of interpolation.

**Comment by underhood** [comment]
@vlvkobal of course, you are right, I confirm other inbound connections do not have this issue

**Comment by underhood** [comment]
#8580 is a problem here also. Much bigger than elsewhere as there are many dimensions being added during runtime

**Comment by underhood** [comment]
As you can see after run of few seconds the charts become dominated by black dimensions that are impossible to tell apart from each other.
Screenshot from 2020-07-29 14-20-23

**Comment by underhood** [comment]
@jacekkolasa Do you have any idea how hard this would be to fix?

**Comment by vlvkobal** [comment]
I think Network Viewer should be disabled by default. We always disable charts with many dimensions. It also will keep in the database old dimension metadata, on every Netdata restart memory usage will grow (@mfundul is it correct?).

@dim08 @ilyam8 - your comments are welcome.

**Comment by jacekkolasa** [comment]
@underhood i'm not sure yet, but I'll take a look. Thanks for linking to the "new dimension" issue

**Comment by vlvkobal** [comment]
Network Viewer charts should have a separate family.

**Comment by ilyam8** [comment]
> your comments are welcome.

Well, we have a bunch of issues because of metdata log + collectors that procude a lot of ephemeral metrics (cgroups was the reason so far). There is no workaround if we dont count disabling cgroups/switch to memore mode save.

Considerin all this i think it makes sense to disable it for now. @mfundul this collector can produce up to 400 (or more, depends on configuration) new metrics on every restart. What do you think?

**Comment by mfundul** [comment]
> 
> 
> > your comments are welcome.
> 
> Well, we have a bunch of issues because of metdata log + collectors that procude a lot of ephemeral metrics (cgroups was the reason so far). There is no workaround if we dont count disabling cgroups/switch to memore mode save.
> 
> Considerin all this i think it makes sense to disable it for now. @mfundul this collector can produce up to 400 (or more, depends on configuration) new metrics on every restart. What do you think?

If those 400 metrics are indeed ephemeral then this is a problem for the metadata footprint, and as a result a problem for agent start-up. It's also a problem for memory footprint because those dimensions will be kept in memory.

**Comment by thiagoftsm** [comment]
> The previous connection that was classified at first as outbound was correctly reclassified and a new connection added. Now I can see traffic only for the second ssh connection.

This was side effect of the changes we made on #9549, I fixed with the last commit.

**Comment by ilyam8** [comment]
> If those 400 metrics are indeed ephemeral

Why do i say that. Dimension is a combination of `ip` and `port`, it is clear that number of such combinations is huge. `Network viewer` adds to charts **first** N combinations, so it is likely we will have different set on every restart. New 400 on restart is worst case scenario, but it doesnt really matter because if it less (real case) it only makes _memory leak_ slower.

**Comment by thiagoftsm** [comment]
Reopening!

**Comment by vlvkobal** [comment]
`ebpf.outbound_bytes` and `ebpf.outbound_bytes` charts have twice as many dimensions as `maximum dimensions`.

**Comment by thiagoftsm** [comment]
>     2. wouldn't it be better for these chart dimensions to be stacked (instead of they being over each other)?
Good suggestion, I brought this in the last commit.

**Comment by thiagoftsm** [comment]
> `ebpf.outbound_bytes` and `ebpf.outbound_bytes` charts have twice as many dimensions as `maximum dimensions`.

This is because we are using `sent`/`recv` to classify traffic, I thought it makes more sense to consider `dimension pair` instead to specify 50 and to have only 25.

@amoss and @dim08 what is your opinion about this?

**Comment by thiagoftsm** [comment]
> as you can see until refresh many dimensions are red histogram line with gray text and in the plot, they are all actually black.
> Although this is not problem of this PR directly I consider this a problem for the Network Viewer as a whole.

This is dashboard issue, our pallete has only 20 colors and when they there is the superposition everything becomes black.

**Comment by underhood** [comment]
@thiagoftsm yes is dashboard issue but is also happening if you have 2 dimensions only and you add new dimensions. So it is not related to 20 limit

**Comment by ilyam8** [comment]
>  if you have 2 dimensions only and you add new dimensions (even if total dimensions in chart is only 3) they will be black already.

This is strange, i didnt see such issue. We have bunch of collecotrs (py and go) that add dimensions in runtime.

> when they there is the superposition everything becomes black.

What is `superposition`? If there is 20+ dimensions we have dimensions with same colors (not black).

**Comment by vlvkobal** [comment]
You made _retransmit chart stacked, but not _bytes and _packets.

**Comment by underhood** [comment]
@ilyam8 I noticed it when I create a custom collector some time ago and created issue #8580. Not sure this bug exists all the time or since React Dasboard etc.

**Comment by thiagoftsm** [comment]
> stacked

Fixed with last commit.

**Comment by ilyam8** [comment]
@underhood, yes, you are right

Screenshot 2020-07-29 at 19 09 22


> since React Dasboard etc.

I think yes, it wasnt like that before react dashboard @jacekkolasa

**Comment by vlvkobal** [comment]
Now I can see data only for the first ssh connection.

ssh-1 (started before Netdata, classified as an outbound connection, you can see traffic on the charts):
image
ssh-2 started (you can see dimensions for a new inbound connection but there is no traffic displayed):
image
ssh-1, typing in a terminal:
image

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Comment by underhood** [comment]
Hi, as per discussion on Slack. There is issue that existing connection is not detected when netdata is started.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5897622&bid=19310201)

**Pull request merged**